### PR TITLE
update validate bookmark.

### PR DESCRIPTION
**Pull request by vwxyzh**
@chenkennt @hellosnow @ansyral @superyyrrzz @qinezh

### CODE REVIEW

File: src/Microsoft.DocAsCode.Build.Common/ValidateBookmark.cs [hunk 1]
```diff
@@ -5,83 +5,49 @@ namespace Microsoft.DocAsCode.Build.Common
...
+    public class ValidateBookmark : IHtmlDocumentHandler
     {
         private static readonly string XpathTemplate = "//*/@{0}";
         private static readonly HashSet<string> WhiteList = new HashSet<string> { "top" };
+        private Dictionary<string, HashSet<string>> registeredBookmarks;
+        private Dictionary<string, List<Tuple<string, string>>> bookmarks;
+        private Dictionary<string, string> fileMapping;
 
-        public ImmutableDictionary<string, object> PrepareMetadata(ImmutableDictionary<string, object> metadata)
+        #region IHtml
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.Common/ValidateBookmark.cs]: >IHtml [](start = 16, length = 5)

minor: IHtmlDocumentHandler Members ?

File: src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs [hunk 2]
```diff
@@ -19,6 +19,7 @@ namespace Microsoft.DocAsCode.Build.Engine
     using Microsoft.DocAsCode.Exceptions;
     using Microsoft.DocAsCode.Plugins;
     using Microsoft.DocAsCode.Utility;
+    using Common;
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.Engine/DocumentBuilder.cs]: >Common [](start = 10, length = 6)

Microsoft.DocAsCode.Build.Common

**Comment by superyyrrzz** [comment]
>         private static ValidateBookmark _validator = new ValidateBookmark();

remove?

---
Refers to: test/Microsoft.DocAsCode.Build.Engine.Tests/ValidateBookmarkTest.cs:20 in eb68a31. [](commit_id = eb68a31dcbd470b74f775174979842b4b5dd3242, deletion_comment = False)

File: src/Microsoft.DocAsCode.Build.Common/ValidateBookmark.cs [hunk 3]
```diff
@@ -5,83 +5,49 @@ namespace Microsoft.DocAsCode.Build.Common
...
+        private Dictionary<string, List<Tuple<string, string>>> bookmarks;
+        private Dictionary<string, string> fileMapping;
 
-        public ImmutableDictionary<string, object> PrepareMetadata(ImmutableDictionary<string, object> metadata)
+        #region IHtmlDocumentHandler members
+
+        public Manifest PreHandle(Manifest manifest)
         {
-            return metadata;
+            registeredBookmarks = new Dictionary<string, HashSet<string>>(new FilePathComparer());
```
Comment (ansyral) [inline src/Microsoft.DocAsCode.Build.Common/ValidateBookmark.cs]: >FilePathComparer [](start = 78, length = 16)

use OSPlatformSensitiveStringComparer

**Comment by ansyral** [comment]
:shipit:

**Pull request merged**